torch/_dynamo/output_graph.py

```diff
@@ -953,6 +990,46 @@ def create_proxy(
         type_expr=None,
         proxy_factory_fn=None,
     ):
+        # NOTE: [Nested SubgraphTracer and free_variable handling]
+        # --------------------------------------------------------
+        # Read NOTE [HigherOrderOperator tracing design] first.
+        #
+        # Let's say we're in the middle of introspecting the body of a possibly
+        # nested HigherOrderOperator, and we see a free variable.
+        #
+        # There are two cases:
+        # 1. We see a free variable that is already tracked by Dynamo.
+        # 2. We see a free variable that has not been tracked by Dynamo
+        #
+        # In case 1, we call `lift_tracked_freevar_to_input` (below)
+        # which will lift the freevar to be an input of this subgraph
+        # and also recursively lift it to be an input on the parent(s).
+        #
+        # In case 2, before the call to `create_proxy`, the InstructionTranslator
+        # will see the freevar when it gets loaded by Python bytecode.
+        # E.g. for Python 3.11 the bytecodes that may do this are LOAD_DEREF or
+        # LOAD_GLOBAL.
+        # There, the InstructionTranslator asks Dynamo to begin tracking the
+        # freevar by building a new Variable.
+        # Building a new Variable automatically lifts the freevar to be an
+        # input of the root SubgraphTracer.
+        #
+        # After that happens, since the variable is now being tracked,
+        # we are now back to case 1.
```

Comment:
Confirming my understanding - is this note effectively saying that:

(1) we will always be in case 1 by the time we get to this code.

(2) Any "free variable" we encounter is guaranteed to already have been lifted into a graph input of the root graph. The additional work we need to do is *only* that we need to lift this free variable into inputs (recursively) of every nested higher-order-op subgraph that we're currently inside of, but not the root graph.

Reply:
Yes! I'll make that clearer